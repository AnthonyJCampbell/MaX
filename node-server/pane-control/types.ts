// Copyright 2021 Metaswitch Networks - Highly Confidential Material
/**
 * Types used just within Pane Control
 *
 * This is to allow some bang-for-buck TypeScript to be written, but should be reconsidered when
 * node-server is fully rewritten in TypeScript.
 */
import { WindowTypes } from "src/shared/types";
export interface CallWindowMetadata {
  remoteParty: string;
}
export interface CallWindowIdentifiers {
  type: WindowTypes;
  uid: string;
}

export interface CallWindow extends Electron.BrowserWindow {
  identifiers: CallWindowIdentifiers;
  metadata: CallWindowMetadata;
}
