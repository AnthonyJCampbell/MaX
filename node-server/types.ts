// Copyright 2021 Metaswitch Networks - Highly Confidential Material
/**
 * Node server types
 *
 * This file contains types for use throughout the Node code, i.e. anything in `node-server/`. It
 * mostly provides methods that convert between these types and protobuf types/binary.
 *
 * For more information on the type system in general, see docs/typescript.md
 */

import {
  Account,
  Analytic,
  CoreAction,
  CoreMotion,
  SettingsAction,
  Contact,
  ActiveCall,
  HistoricCall,
  ContactsAction,
  ContactsMotion,
  Chat,
  ChatMessage,
  Meeting,
  MeetingAction,
  MessagingAction,
  WispaMessage,
  DeepPartial,
  Settings,
  User,
  UpdateableUser,
  VoicemailFaxCount,
  VoicemailFaxAction,
} from "../src/shared/types";
import { Result } from "neverthrow";

/**
 * Any type which can be sent over WISPA by itself (as opposed to as part of another type)
 */
export type WispaDataType =
  | Account
  | CoreAction
  | CoreMotion
  | Settings
  | SettingsAction
  | ContactsAction
  | Contact
  | ContactsMotion
  | ActiveCall
  | HistoricCall
  | Chat
  | ChatMessage
  | User
  | UpdateableUser
  | Meeting
  | MeetingAction
  | MessagingAction
  | VoicemailFaxCount
  | VoicemailFaxAction
  | Analytic;

/**
 * All types of message that can be sent to Wispa
 */
export enum WispaRequestMethod {
  /**
   * Request a specific piece of Wispa data, e.g. the Contact with UID "1234"
   */
  get = "get",

  /**
   * Request a list of all data in a given namespace, e.g. all Contacts
   */
  list = "list",

  /**
   * Create a new instance of a Wispa data type
   */
  create = "create",

  /**
   * Update an existing piece of Wispa data with new information
   */
  update = "update",

  /**
   * Delete the specified Wispa data
   */
  delete = "delete",

  /**
   * Request a specified action
   */
  action = "action",
}

/**
 * All types of message that can be received from Wispa
 */
export enum WispaMessageMethod {
  /**
   * A single piece of Wispa data
   */
  data = "data",

  /**
   * A list of pieces of Wispa data
   */
  datalist = "datalist",

  /**
   * A piece of Wispa data that has been deleted
   */
  deleted = "deleted",

  /**
   * Request a specified motion
   */
  motion = "motion",
}

export enum Namespace {
  settings = "settings",
  contacts = "contacts",
  activecalls = "activecalls",
  callhistory = "callhistory",
  messaging = "messaging",
  meetings = "meetings",
  core = "core",
  voicemails = "voicemails",
  analytics = "analytics",
  user = "user",
}

/**
 * All errors that occurred during decoding of some protobuf binary into WispaData objects
 */
export interface DecodeErrors {
  // Text identifying the object that failed to decode, e.g. its UID
  identifier: string;

  // Array of strings each describing an error that occurred during decoding
  errorTexts: string[];
}

/**
 * Required methods to handle receiving a given WispaDataType from Wispa
 */
export interface WispaReceiver<T extends WispaDataType> {
  decode: (data: Uint8Array, defaultMissingFields?: boolean) => Result<T, DecodeErrors>;
  decodeList: (data: Uint8Array, defaultMissingFields?: boolean) => Result<T, DecodeErrors>[];
  buildMessage: (options: { data?: T; dataList?: T[]; deleted?: T; motion?: T }) => WispaMessage;
}

/**
 * Required methods to handle sending a given WispaDataType to Wispa
 */
export interface WispaSender<T extends WispaDataType> {
  encode: (data: DeepPartial<T>) => Uint8Array;
}

/**
 * The source of truth for all the ports wispa could be running on in various situations
 */
export const wispaPorts = {
  // The official installed Alpha client
  PRODUCTION: "9090",

  // Electron FVs
  ELECTRON_FV: "9091",

  // The default port. Used during development, e.g. "npm start" in this codebase or "./run.sh"
  // in AccessionDesktop.
  DEVELOPMENT: "9092",

  // UC Daemon FVs - not used by this codebase but included for completeness
  UC_DAEMON_FV: "9093",

  // The Early Access client (temporary ~April 2021)
  EARLY_ACCESS: "9094",
};

// Re-export shared types. This allows node-server code to import all its types from one file (i.e.
// this one). The above definitions take precedence in the event of any clashes
export * from "../src/shared/types";
