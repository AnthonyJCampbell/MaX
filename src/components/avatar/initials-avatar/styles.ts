// Copyright 2021 Metaswitch Networks - Highly Confidential Material
import styled from "styled-components";
import { Colors as C, Sizes as S, zIndexes } from "components/utils/style-constants";

const accountInitialsElementsSize = "40px";
const contactInitialsElementsSize = "32px";

interface HasSidebarProp {
  isSidebar: boolean;
}

export const InitialsAvatarBackground = styled.div<HasSidebarProp>`
  ${({ isSidebar }): string =>
    isSidebar
      ? `
  width: ${accountInitialsElementsSize};
  height: ${accountInitialsElementsSize};`
      : `
  width: ${contactInitialsElementsSize};
  height: ${contactInitialsElementsSize};`}

  background-color: ${C.midbar.defaultAvatarBackground};
  border: 1px solid white;

  object-fit: cover;
  border-radius: 50%;
  vertical-align: -100%;

  z-index: ${zIndexes.avatarBackground};
  position: relative;
`;

export const AvatarText = styled.p<HasSidebarProp>`
  ${({ isSidebar }): string =>
    isSidebar
      ? `
  font-size: ${S.avatar.accountInitialsFontSize};
  line-height: ${accountInitialsElementsSize};
  width: ${accountInitialsElementsSize};
  height: ${accountInitialsElementsSize};`
      : `
  font-size: ${S.avatar.contactInitialsFontSize};
  line-height: ${contactInitialsElementsSize};
  width: ${contactInitialsElementsSize};
  height: ${contactInitialsElementsSize};`}

  color: ${C.midbar.defaultAvatarForeground};
  text-align: center;
  font-weight: lighter;
`;
