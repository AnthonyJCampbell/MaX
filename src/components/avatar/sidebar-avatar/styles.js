// Copyright 2021 Metaswitch Networks - Highly Confidential Material
import styled from "styled-components";
import {
  Colors as C,
  Effects as E,
  primaryColour,
  zIndexes,
} from "components/utils/style-constants";

export const SideBarPictureContainer = styled.button`
  width: 50px;
  height: 50px;
  border: ${C.transparentBorder};
  display: flex;
  border-radius: 50%;
  padding: 2px;

  transition: ${E.transition};
  &:focus {
    border: ${C.whiteFocusBorder};
    box-shadow: ${C.whiteFocusShadow};
  }
  &:hover {
    cursor: pointer;
    box-shadow: ${C.whiteFocusShadow};
    border: ${C.whiteHoverBorder};
  }
  &:active {
    cursor: pointer;
    border: ${C.whiteFocusBorder};
    box-shadow: ${C.whiteFocusShadow};
    opacity: 0.7;
  }
`;

export const SidebarPresenceIcon = styled.img`
  width: 13px;
  height: 13px;
`;

export const SidebarPresenceBorder = styled.div`
  width: 13px;
  height: 13px;

  position: relative;
  top: 28px;
  right: 15px;
  z-index: 2;

  border-radius: 50%;
  border: 2px solid ${primaryColour};
  background-color: ${primaryColour};
`;

export const SidebarPicture = styled.img`
  width: 42px;
  height: 42px;
  border-radius: 50%;
`;

export const SidebarCustomStatusIcon = styled.img`
  width: 12px;
  height: 12px;
`;

export const SidebarCustomStatusBorder = styled.div`
  width: 12px;
  height: 12px;

  position: relative;
  top: 16px;
  right: 24px;
  z-index: 3;

  border-radius: 50%;
  border: 2px solid ${primaryColour};
  background-color: ${primaryColour};
`;

export const SidebarPresencePlaceholder = styled.div`
  min-width: 15px;
  min-height: 15px;
  object-fit: cover;
  border-radius: 50%;
  border: ${C.transparentBorder};
  margin-right: -5px;
  vertical-align: -110%;
  z-index: ${zIndexes.avatarPresenceIcon};
  position: relative;
  top: 28px;
  right: 12px;
`;
