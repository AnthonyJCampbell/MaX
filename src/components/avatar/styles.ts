// Copyright 2021 Metaswitch Networks - Highly Confidential Material
import styled from "styled-components";
import { zIndexes } from "../utils/style-constants";

export const ImageContainer = styled.div`
  position: relative;
  display: flex;
  margin-right: -2px;
`;

export const ProfilePicture = styled.img`
  width: 34px;
  height: 34px;

  object-fit: cover;
  border-radius: 50%;
  vertical-align: -110%;
  z-index: ${zIndexes.avatarBackground}
  position: relative;
`;

export const PresenceImage = styled.img`
  width: 12px;
  height: 12px;
  object-fit: cover;
  border-radius: 50%;

  margin-right: -5px;
  vertical-align: -110%;
  z-index: ${zIndexes.avatarPresenceIcon};
  position: relative;
  top: 23px;
  right: 10px;
`;

export const PresencePlaceholder = styled.div`
  padding-right: 7px;
`;
