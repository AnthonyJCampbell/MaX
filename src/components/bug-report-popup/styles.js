// Copyright 2021 Metaswitch Networks - Highly Confidential Material
import styled, { keyframes } from "styled-components";
import {
  primaryColour,
  Colors as C,
  Effects as E,
  zIndexes,
} from "components/utils/style-constants";

const fade = keyframes`
  from {
    opacity: 0;
  }
  to {
    opacity: 1;
  }
`;

export const PopupContainer = styled.span`
  box-sizing: border-box;
  width: 600px;
  height: 240px;

  border-radius: 4px;
  box-shadow: 0px 2px 4px #00000050;
  padding: 4px;

  position: fixed;
  bottom: 6px;
  left: 6px;
  z-index: ${zIndexes.unimplementedPopup};

  background: white;

  animation: ${fade} 0.5s linear;

  display: flex;
`;

export const ImageContainer = styled.span`
  box-sizing: border-box;
  height: 100%;
`;
export const Image = styled.img`
  box-sizing: border-box;
  height: 100%;
`;
export const TextContainer = styled.span`
  box-sizing: border-box;
  height: 100%;

  padding: 6px 4px 2px 4px;

  display: flex;
  flex-flow: column;

  p,
  li {
    user-select: text;
  }

  button {
    color: ${primaryColour};
    border: none;
    font: inherit;

    padding: 0px;

    transition: ${E.transition};
    &:focus {
      border: ${C.focusBorder};
      box-shadow: ${C.focusShadow};
    }
    &:hover {
      cursor: pointer;
      opacity: ${E.hoverOpacity};
    }
    &:active {
      cursor: pointer;
      opacity: ${E.activeOpacity};
    }
  }
`;

export const ButtonsContainer = styled.span`
  box-sizing: border-box;
  width: 100%;

  padding: 10px 0px 0px 0px;

  display: flex;
  justify-content: space-between;
  button {
    color: ${primaryColour};
    font-weight: 700;
    font-size: 12px;
    border: none;
    padding: 0px;

    transition: ${E.transition};
    &:focus {
      border: ${C.focusBorder};
      box-shadow: ${C.focusShadow};
    }
    &:hover {
      cursor: pointer;
      opacity: ${E.hoverOpacity};
    }
    &:active {
      cursor: pointer;
      opacity: ${E.activeOpacity};
    }
  }
`;

export const DismissButton = styled.span`
  box-sizing: border-box;
  padding-bottom: 10px;
`;
