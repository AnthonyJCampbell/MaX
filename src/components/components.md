# Components

The entire app (`mainWindow` as defined by Electron) is primarily made up of three panes that contain the majority of the functionality.

- The left-hand `Sidebar`, which contains the buttons to switch between different tabs on the `Midbar`. This pane is the primary driver of user behaviour.
- The `Midbar` is a scrollable list containing the contacts/call history/chat messages/meeting controls, which can change the contents of `Rightbar` when clicked.
- The `Rightbar` is the primary information pane. It shows the contact details for the contact selected in `Midbar`.

`Avatar` is a component shared between the three main panes.
