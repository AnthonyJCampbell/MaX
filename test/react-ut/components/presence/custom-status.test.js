// Copyright 2021 Metaswitch Networks - Highly Confidential Material
import "test/react-ut/components/require-mock";
import React from "react";
import { shallow, configure } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import * as redux from "react-redux";
import renderer from "react-test-renderer";
import "jest-styled-components";

import { CustomStatusState } from "components/utils/common";
import CustomStatus, { MAX_CUSTOM_STATUS_LENGTH } from "components/presence/custom-status";
import { PresenceState } from "shared/types";
import { Button } from "components/menu/menu";
import * as S from "components/presence/styles";
import { mockedT } from "shared/mocks/ts-utils";
import log from "src/renderer-logging";

configure({ adapter: new Adapter() });

jest.mock("react-redux");
jest.mock("store/store");
jest.mock("src/renderer-logging");
jest.mock("react", () => {
  const react = jest.requireActual("react");

  return { ...react, useRef: jest.fn() };
});

const events = {
  onSave: jest.fn(),
  onCancel: jest.fn(),
  onClose: jest.fn(),
  onEdit: jest.fn(),
  onClean: jest.fn(),
  onClick: jest.fn(),
};

beforeEach(() => log.test(`Start running: ${expect.getState().currentTestName}`));

afterEach(() => {
  jest.clearAllMocks();
  log.test(`End running: ${expect.getState().currentTestName}`);
});

const useSelectorMock = jest.spyOn(redux, "useSelector");

const mockReducerState = (override) => {
  const mockState = {
    paneManagementReducer: {
      customStatusState: override?.customStatusState || CustomStatusState.INITIAL,
    },
    userReducer: {
      user: {
        presence: {
          state: PresenceState.ONLINE,
          customStatus: override?.customStatus || "",
        },
      },
    },
  };

  useSelectorMock.mockImplementation((callback) => callback(mockState));
};

const customStatus = "An custom status";

const testMenuItemButtonScenario = () => {
  // Snapshot the HTML and CSS
  const componentJson = renderer.create(<CustomStatus id="anId" events={events} />).toJSON();
  expect(componentJson).toMatchSnapshot();

  const wrap = shallow(<CustomStatus id="anId" events={events} />);

  const menuItemButton = wrap.find(Button);
  expect(menuItemButton.exists()).toBe(true);
  expect(menuItemButton.prop("text")).toEqual(mockedT("setCustomStatus"));
};

const testNonEditingScenario = () => {
  // Snapshot the HTML and CSS
  const componentJson = renderer.create(<CustomStatus id="anId" events={events} />).toJSON();
  expect(componentJson).toMatchSnapshot();

  const wrap = shallow(<CustomStatus id="anId" events={events} />);

  const menuItemButton = wrap.find(Button);
  expect(menuItemButton.exists()).toBe(false);

  const CustomStatusContainer = wrap.find(S.CustomStatusContainer);
  expect(CustomStatusContainer.exists()).toBe(true);

  const customStatusParagraph = wrap.find(S.CustomStatusParagraph);
  expect(customStatusParagraph.exists()).toBe(true);
  expect(customStatusParagraph.text()).toEqual(customStatus);

  const actionButtons = wrap.find(S.ActionButton);
  expect(actionButtons).toHaveLength(2);

  const editButton = actionButtons.at(0);
  expect(editButton.text()).toEqual(mockedT("edit"));
  editButton.simulate("click");
  expect(events.onEdit).toHaveBeenCalled();

  const clearButton = actionButtons.at(1);
  expect(clearButton.text()).toEqual(mockedT("clear"));
  clearButton.simulate("click");
  expect(events.onClean).toHaveBeenCalled();
};

const testEditingScenario = (expectedTextAreaValue) => {
  // Snapshot the HTML and CSS
  const componentJson = renderer.create(<CustomStatus id="anId" events={events} />).toJSON();
  expect(componentJson).toMatchSnapshot();

  const wrap = shallow(<CustomStatus id="anId" events={events} />);

  const menuItemButton = wrap.find(Button);
  expect(menuItemButton.exists()).toBe(false);

  const CustomStatusContainer = wrap.find(S.CustomStatusContainer);
  expect(CustomStatusContainer.exists()).toBe(true);

  const textArea = wrap.find(S.TextArea);
  expect(textArea.exists()).toBe(true);
  expect(textArea.prop("value")).toEqual(expectedTextAreaValue);

  const actionButtons = wrap.find(S.ActionButton);
  expect(actionButtons).toHaveLength(2);

  const saveButton = actionButtons.at(0);
  expect(saveButton.text()).toEqual(mockedT("save"));
  saveButton.simulate("click");
  expect(events.onSave).toHaveBeenCalledWith(expectedTextAreaValue);

  const cancelButton = actionButtons.at(1);
  expect(cancelButton.text()).toEqual(mockedT("cancel"));
  cancelButton.simulate("click");
  expect(events.onCancel).toHaveBeenCalledTimes(1);

  const characterCount = wrap.find(S.CharacterCount);
  expect(characterCount.exists()).toBe(true);
  const expectedCharacterCount = MAX_CUSTOM_STATUS_LENGTH - expectedTextAreaValue?.length;
  expect(characterCount.text()).toEqual(`${expectedCharacterCount}`);
};

describe("<CustomStatus />", () => {
  const useDispatchMock = jest.spyOn(redux, "useDispatch");
  useDispatchMock.mockReturnValue(jest.fn());
  const stopPropagation = jest.fn();
  const preventDefault = jest.fn();

  /* eslint-disable-next-line jest/expect-expect */
  it("Renders in the initial state", () => {
    mockReducerState();

    testMenuItemButtonScenario();
  });

  /* eslint-disable-next-line jest/expect-expect */
  it("Renders in initial state with a custom status already set", () => {
    mockReducerState({
      customStatus,
    });

    testNonEditingScenario();
  });

  /* eslint-disable-next-line jest/expect-expect */
  it("Renders in the editing state", () => {
    mockReducerState({
      customStatusState: CustomStatusState.EDITING,
    });

    testEditingScenario("");
  });

  /* eslint-disable-next-line jest/expect-expect */
  it("Renders in the editing state with a custom status already set", () => {
    mockReducerState({
      customStatusState: CustomStatusState.EDITING,
      customStatus,
    });

    testEditingScenario(customStatus);
  });

  /* eslint-disable-next-line jest/expect-expect */
  it("Renders in the close state", () => {
    mockReducerState({
      customStatusState: CustomStatusState.CLOSE,
    });

    testMenuItemButtonScenario();
  });

  /* eslint-disable-next-line jest/expect-expect */
  it("Renders in the close state with a custom status already set", () => {
    mockReducerState({
      customStatusState: CustomStatusState.CLOSE,
      customStatus,
    });

    testNonEditingScenario();
  });

  it("Should call onCancel when Esc is pressed in Editing Mode", () => {
    mockReducerState({
      customStatusState: CustomStatusState.EDITING,
    });

    const wrap = shallow(<CustomStatus id="anId" events={events} />);

    wrap.find(S.CustomStatusContainer).simulate("keydown", { key: "Escape", stopPropagation });
    expect(stopPropagation).toHaveBeenCalled();
    expect(events.onCancel).toHaveBeenCalled();
  });

  it("Should call onClose when Esc is pressed in Initial Mode", () => {
    mockReducerState({
      customStatusState: CustomStatusState.INITIAL,
      customStatus,
    });

    const wrap = shallow(<CustomStatus id="anId" events={events} />);

    wrap.find(S.CustomStatusContainer).simulate("keydown", { key: "Escape", stopPropagation });
    expect(stopPropagation).toHaveBeenCalled();
    expect(events.onClose).toHaveBeenCalled();
  });

  it("Should call onSave when Enter is pressed in Editing Mode", () => {
    mockReducerState({
      customStatusState: CustomStatusState.EDITING,
      customStatus,
    });

    const useRefSpy = jest.spyOn(React, "useRef").mockReturnValueOnce({ current: true });

    const wrap = shallow(<CustomStatus id="anId" events={events} />);
    wrap.simulate("focus");
    wrap.find(S.CustomStatusContainer).simulate("keydown", {
      key: "Enter",
      stopPropagation,
      preventDefault,
    });
    expect(useRefSpy).toHaveBeenCalled();
    expect(stopPropagation).toHaveBeenCalled();
    expect(preventDefault).toHaveBeenCalled();
    expect(events.onSave).toHaveBeenCalledWith(customStatus);
  });
});
