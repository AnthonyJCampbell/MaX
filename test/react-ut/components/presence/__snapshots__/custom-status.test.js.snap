// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CustomStatus /> Renders in initial state with a custom status already set 1`] = `
.c0 {
  background-color: #f7f7f7;
  margin-left: 6px;
  margin-top: 4px;
  position: relative;
  max-width: 178px;
}

.c1 {
  position: absolute;
  top: 8px;
  left: 7px;
}

.c2 {
  background-color: #f7f7f7;
  box-sizing: border-box;
  display: block;
  font-size: 12px;
  height: auto;
  line-height: 1.2;
  max-height: 64px;
  margin-bottom: 7px;
  padding-left: 28px;
  padding-top: 8px;
  overflow-y: auto;
  word-break: break-word;
  width: 99%;
}

.c3 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 5px 1px 0 1px;
  background: #ffffff;
}

.c4 {
  border: 2px solid transparent;
  color: #2F78CA;
  font-weight: 700;
  font-size: 12px;
  padding: 2px 4px;
  -webkit-transition: 0.2s;
  transition: 0.2s;
}

.c4:first-child {
  margin-right: 2px;
}

.c4:focus {
  border: 2px solid #2F78CA;
  box-shadow: inset 0 0 0 2px white;
}

.c4:hover {
  cursor: pointer;
  border: 2px solid #2F78CA;
}

.c4:active {
  cursor: pointer;
}

.c4:disabled {
  cursor: default;
  opacity: 0.6;
  border: 2px solid transparent;
}

<div
  className="c0"
  id="anId"
  onKeyDown={[Function]}
>
  <img
    alt=""
    className="c1"
    src="menu-status"
  />
  <p
    className="c2"
    tabIndex={0}
  >
    An custom status
  </p>
  <div
    className="c3"
  >
    <button
      className="c4"
      id="customStatusButtonEdit"
      onClick={[MockFunction]}
    >
      edit
    </button>
    <button
      className="c4"
      id="customStatusButtonClear"
      onClick={[Function]}
    >
      clear
    </button>
  </div>
</div>
`;

exports[`<CustomStatus /> Renders in the close state 1`] = `
.c0 {
  padding: 4px 12px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  min-width: 190px;
  width: 100%;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: left;
  -webkit-justify-content: left;
  -ms-flex-pack: left;
  justify-content: left;
  border: none;
  width: 100%;
  font-size: 12px;
  opacity: 1;
  -webkit-transition: 0.2s;
  transition: 0.2s;
}

.c0:focus {
  background-color: #2F78CA4d;
}

.c0:hover {
  cursor: pointer;
  background-color: #2F78CA1a;
  -webkit-transition: 0s;
  transition: 0s;
}

.c0:active {
  cursor: pointer;
  background-color: #2F78CA80;
}

.c0 p {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c0 img {
  margin-right: 8px;
  padding-top: 1px;
  max-width: 14px;
  max-height: 14px;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  color: #000000de;
}

.c2 {
  color: #000000;
}

.c3 {
  color: #000000a6;
  margin-left: 3px;
  display: inline-block;
}

<button
  className="c0"
  disabled={false}
  id="anId"
  onClick={[Function]}
>
  <img
    alt=""
    src="menu-status"
  />
  <div
    className="c1"
  >
    <p
      className="c2"
    >
      setCustomStatus
    </p>
    <p
      className="c3"
    />
  </div>
</button>
`;

exports[`<CustomStatus /> Renders in the close state with a custom status already set 1`] = `
.c0 {
  background-color: #f7f7f7;
  margin-left: 6px;
  margin-top: 4px;
  position: relative;
  max-width: 178px;
}

.c1 {
  position: absolute;
  top: 8px;
  left: 7px;
}

.c2 {
  background-color: #f7f7f7;
  box-sizing: border-box;
  display: block;
  font-size: 12px;
  height: auto;
  line-height: 1.2;
  max-height: 64px;
  margin-bottom: 7px;
  padding-left: 28px;
  padding-top: 8px;
  overflow-y: auto;
  word-break: break-word;
  width: 99%;
}

.c3 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 5px 1px 0 1px;
  background: #ffffff;
}

.c4 {
  border: 2px solid transparent;
  color: #2F78CA;
  font-weight: 700;
  font-size: 12px;
  padding: 2px 4px;
  -webkit-transition: 0.2s;
  transition: 0.2s;
}

.c4:first-child {
  margin-right: 2px;
}

.c4:focus {
  border: 2px solid #2F78CA;
  box-shadow: inset 0 0 0 2px white;
}

.c4:hover {
  cursor: pointer;
  border: 2px solid #2F78CA;
}

.c4:active {
  cursor: pointer;
}

.c4:disabled {
  cursor: default;
  opacity: 0.6;
  border: 2px solid transparent;
}

<div
  className="c0"
  id="anId"
  onKeyDown={[Function]}
>
  <img
    alt=""
    className="c1"
    src="menu-status"
  />
  <p
    className="c2"
    tabIndex={0}
  >
    An custom status
  </p>
  <div
    className="c3"
  >
    <button
      className="c4"
      id="customStatusButtonEdit"
      onClick={[MockFunction]}
    >
      edit
    </button>
    <button
      className="c4"
      id="customStatusButtonClear"
      onClick={[Function]}
    >
      clear
    </button>
  </div>
</div>
`;

exports[`<CustomStatus /> Renders in the editing state 1`] = `
.c0 {
  background-color: #f7f7f7;
  margin-left: 6px;
  margin-top: 4px;
  position: relative;
  max-width: 178px;
}

.c1 {
  position: absolute;
  top: 8px;
  left: 7px;
}

.c2 {
  background-color: #f7f7f7;
  box-sizing: border-box;
  border: none;
  font-size: 12px;
  height: 45px;
  line-height: 1.2;
  margin-top: 4px;
  padding-left: 28px;
  resize: none;
  width: 99%;
}

.c3 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 5px 1px 0 1px;
  background: #ffffff;
}

.c4 {
  border: 2px solid transparent;
  color: #2F78CA;
  font-weight: 700;
  font-size: 12px;
  padding: 2px 4px;
  -webkit-transition: 0.2s;
  transition: 0.2s;
}

.c4:first-child {
  margin-right: 2px;
}

.c4:focus {
  border: 2px solid #2F78CA;
  box-shadow: inset 0 0 0 2px white;
}

.c4:hover {
  cursor: pointer;
  border: 2px solid #2F78CA;
}

.c4:active {
  cursor: pointer;
}

.c4:disabled {
  cursor: default;
  opacity: 0.6;
  border: 2px solid transparent;
}

.c5 {
  font-size: 12px;
  font-weight: 600;
  margin-left: auto;
  margin-right: 8px;
  text-align: right;
  width: 24px;
  color: #636363;
}

<div
  className="c0"
  id="anId"
  onKeyDown={[Function]}
>
  <img
    alt=""
    className="c1"
    src="menu-status"
  />
  <textarea
    aria-label="customStatusTextDefault"
    aria-live="polite"
    className="c2"
    id="customStatus"
    name="customStatus"
    onChange={[Function]}
    value=""
  />
  <div
    className="c3"
  >
    <button
      className="c4"
      disabled={false}
      id="customStatusButtonSave"
      onClick={[Function]}
    >
      save
    </button>
    <button
      className="c4"
      id="customStatusButtonCancel"
      onClick={[Function]}
    >
      cancel
    </button>
    <span
      className="c5"
    >
      100
    </span>
  </div>
</div>
`;

exports[`<CustomStatus /> Renders in the editing state with a custom status already set 1`] = `
.c0 {
  background-color: #f7f7f7;
  margin-left: 6px;
  margin-top: 4px;
  position: relative;
  max-width: 178px;
}

.c1 {
  position: absolute;
  top: 8px;
  left: 7px;
}

.c2 {
  background-color: #f7f7f7;
  box-sizing: border-box;
  border: none;
  font-size: 12px;
  height: 45px;
  line-height: 1.2;
  margin-top: 4px;
  padding-left: 28px;
  resize: none;
  width: 99%;
}

.c3 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 5px 1px 0 1px;
  background: #ffffff;
}

.c4 {
  border: 2px solid transparent;
  color: #2F78CA;
  font-weight: 700;
  font-size: 12px;
  padding: 2px 4px;
  -webkit-transition: 0.2s;
  transition: 0.2s;
}

.c4:first-child {
  margin-right: 2px;
}

.c4:focus {
  border: 2px solid #2F78CA;
  box-shadow: inset 0 0 0 2px white;
}

.c4:hover {
  cursor: pointer;
  border: 2px solid #2F78CA;
}

.c4:active {
  cursor: pointer;
}

.c4:disabled {
  cursor: default;
  opacity: 0.6;
  border: 2px solid transparent;
}

.c5 {
  font-size: 12px;
  font-weight: 600;
  margin-left: auto;
  margin-right: 8px;
  text-align: right;
  width: 24px;
  color: #636363;
}

<div
  className="c0"
  id="anId"
  onKeyDown={[Function]}
>
  <img
    alt=""
    className="c1"
    src="menu-status"
  />
  <textarea
    aria-label="customStatusText-{\\"customStatusText\\":\\"An custom status\\",\\"remainingCharacters\\":84}"
    aria-live="polite"
    className="c2"
    id="customStatus"
    name="customStatus"
    onChange={[Function]}
    value="An custom status"
  />
  <div
    className="c3"
  >
    <button
      className="c4"
      disabled={false}
      id="customStatusButtonSave"
      onClick={[Function]}
    >
      save
    </button>
    <button
      className="c4"
      id="customStatusButtonCancel"
      onClick={[Function]}
    >
      cancel
    </button>
    <span
      className="c5"
    >
      84
    </span>
  </div>
</div>
`;

exports[`<CustomStatus /> Renders in the initial state 1`] = `
.c0 {
  padding: 4px 12px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  min-width: 190px;
  width: 100%;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: left;
  -webkit-justify-content: left;
  -ms-flex-pack: left;
  justify-content: left;
  border: none;
  width: 100%;
  font-size: 12px;
  opacity: 1;
  -webkit-transition: 0.2s;
  transition: 0.2s;
}

.c0:focus {
  background-color: #2F78CA4d;
}

.c0:hover {
  cursor: pointer;
  background-color: #2F78CA1a;
  -webkit-transition: 0s;
  transition: 0s;
}

.c0:active {
  cursor: pointer;
  background-color: #2F78CA80;
}

.c0 p {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c0 img {
  margin-right: 8px;
  padding-top: 1px;
  max-width: 14px;
  max-height: 14px;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  color: #000000de;
}

.c2 {
  color: #000000;
}

.c3 {
  color: #000000a6;
  margin-left: 3px;
  display: inline-block;
}

<button
  className="c0"
  disabled={false}
  id="anId"
  onClick={[Function]}
>
  <img
    alt=""
    src="menu-status"
  />
  <div
    className="c1"
  >
    <p
      className="c2"
    >
      setCustomStatus
    </p>
    <p
      className="c3"
    />
  </div>
</button>
`;
